### PR TITLE
Implement [BLB] Rottenmouth Viper

### PR DESCRIPTION
**Pull request by Momo2907**
Messed up the commit history a bit, the work on Helga, Skittish Seer is currently on another branch in the forked repo.
Played a couple games with the card against a bot and it seems to be working properly.
The implementation for the CostModificationAbility is based on Dargo, the Shipwrecker.
The implementation for the EntersOrAttacksAbility is based on Torment of Hailfire.

### CODE REVIEW

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        return true;
+    }
+
+    @Override
+    public boolean applies(Ability abilityToModify, Ability source, Game game) {
+        return abilityToModify instanceof SpellAbility && abilityToModify.getSourceId().equals(source.getSourceId());
+    }
+}
+
+class RottenmouthViperCostModificationWatcher extends Watcher {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:176]: seems overimplemented, Dargo needs a watcher but this doesn't, see Torgaar, Famine Incarnate

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                                Permanent permanent = game.getPermanent(target.getFirstTarget());
+                                if (permanent != null) {
+                                    if (permanent.sacrifice(source, game)) {
+                                        continue;
+                                    }
+                                }
+                            }
+                        }
+                        if (!opponent.getHand().isEmpty() && opponent.chooseUse(outcome, "Discard a card? (Iteration " + i + " of " + repeat + ")",
+                                "Otherwise you lose 4 life.", "Discard", "Lose 3 life", source, game)) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:118]: 4 life

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+    public RottenmouthViperEffect(DynamicValue blightCounterAmount) {
+        super(Outcome.LoseLife);
+        this.blightCounterAmount = blightCounterAmount;
+        this.staticText = "Then for each blight counter on it, each opponent loses 4 life unless that player sacrifices a nonland permanent or discards a card.";
+    }
+
+    private RottenmouthViperEffect(final RottenmouthViperEffect effect) {
+        super(effect);
+        this.blightCounterAmount = effect.blightCounterAmount;
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:85]: copy()

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    public RottenmouthViper copy() {
+        return new RottenmouthViper(this);
+    }
+}
+
+class RottenmouthViperEffect extends OneShotEffect{
+
+    DynamicValue blightCounterAmount;
+
+    public RottenmouthViperEffect(DynamicValue blightCounterAmount) {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:77]: internal custom class, no public modifier here

**Review by xenohedron — CHANGES_REQUESTED** [review]

**Review by xenohedron — APPROVED** [review]

**Comment by xenohedron** [comment]
thanks for the contribution

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null){
+            int repeat = blightCounterAmount.calculate(game, source, this);
+            for (int i = 1; i <= repeat; i++) {
+
+                for (UUID opponentId : game.getOpponents(source.getControllerId())) {
+                    Player opponent = game.getPlayer(opponentId);
+                    if (opponent != null) {
+                        int permanents = game.getBattlefield().countAll(StaticFilters.FILTER_PERMANENT_NON_LAND, opponentId, game);
+                        if (permanents > 0 && opponent.chooseUse(outcome, "Sacrifices a nonland permanent? (Iteration " + i + " of " + repeat + ")",
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:98]: AI works bad with default outcome (`Outcome.LoseLife`). It's good practice to use custom outcome for each inner choose.

Must use `Outcome.Benefit` so AI will use sacrifice instead loose life all the time. In ideal world it's better to insert more complex logic like if opponent life > 10 then use else not use, but Outcome.Benefit is minimum.

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                                "Otherwise you have to discard a card or lose 4 life.", "Sacrifice", "Discard or life loss", source, game)) {
+                            Target target = new TargetPermanent(StaticFilters.FILTER_CONTROLLED_PERMANENT_NON_LAND);
+                            target.withNotTarget(true);
+                            if (opponent.choose(outcome, target, source, game)) {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:102]: But not here. It must use bad outcome cause it choose card to sacrifice. So must use `Outcome.Sacrifice`.

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                                Permanent permanent = game.getPermanent(target.getFirstTarget());
+                                if (permanent != null) {
+                                    if (permanent.sacrifice(source, game)) {
+                                        continue;
+                                    }
+                                }
+                            }
+                        }
+                        if (!opponent.getHand().isEmpty() && opponent.chooseUse(outcome, "Discard a card? (Iteration " + i + " of " + repeat + ")",
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/r/RottenmouthViper.java:111]: Same here. Must Outcome.Benefit so AI will discard instead loose life all the time.

**Review by JayDi85 — COMMENTED** [review]
AI related things also important for complex cards

File: Mage.Sets/src/mage/cards/r/RottenmouthViper.java (hunk 5 above)
Comment (JayDi85) [reply Mage.Sets/src/mage/cards/r/RottenmouthViper.java:98]: Added AI hints by 84db421f6dde227d635de7d6a4b28f9a7df1861c

**Pull request merged**